src/ppu/methods/utils.py

```diff
@@ -0,0 +1,221 @@
+import random
+from copy import deepcopy
+
+import numpy as np
+import pandas as pd
+from scipy.special import logsumexp
+from sklearn.datasets import make_moons
+
+
+def weighted_mean(x, **kws):
+    return np.sum(np.real(x) * np.imag(x)) / np.sum(np.imag(x))
+
+def weighted_sd(x, **kws):
+    mu = weighted_mean(x)
+    sd = np.sqrt(np.sum((np.real(x) - mu)**2 * np.imag(x)) / np.sum(np.imag(x)))
+    return mu-sd, mu+sd
+
+def stable_logit_transform(probs, max_=16):
+    with np.errstate(divide = "ignore", invalid="ignore"):
+        probs = np.array(probs)
+        logs = np.log(probs)
+        anti_logs = np.log(np.subtract(1, probs))
+        results = logs - anti_logs
+        return np.clip(results, -max_, max_)
+
+def LSE(z):
+    """input: reduced logits"""
+    #if len(x.shape) < 2:
+    z = np.hstack([z, np.zeros(z.shape)])
+    #x[:,1] = 0
+    return logsumexp(z, axis=1)
+
+def LSE_LB_estimator_(x, C):
+    r""" 
+    LSE lower bound estimator
+    C: number of summands in the estimator - higher is better but slower
+
+    Computes the following formula:
+    \frac{1}{2n} \\sum_{i=1}^n X_i + \\sum_{j=1}^C \\ln (1 + \frac{\frac{1}{n(n-1)} \\sum_{k \neq l} X_k X_l}{4 (n - 0.5)^2 \\pi^2}) + \\ln 2
+    """
+    assert x.shape[1]==1
+    n = x.shape[0]
+    avg = np.mean(x)
+    xxT = np.outer(x, x)
+    # \frac{1}{n(n-1)} \sum_{i \neq j} X_i X_j
+    avg_sq = np.mean(xxT)*n/(n-1) - np.mean(np.diag(xxT))/(n-1)
+    #avg_sq = avg**2 - np.var(x)/n
+    sums = np.sum([np.log1p(avg_sq / (4*np.pi**2*(j+0.5)**2)) for j in range(C)])
+    return avg/2 + sums + np.log(2)
+
+def BI_LSE(z, bound="auto", C=1000):
+    r"""Bregman information generated by LSE(x) = ln(1 + \sum_i e^{x_i})
+    `bound`={'auto', 'lower', 'upper'} sets if upper or lower bound should be estimated;
+    'upper' is only possible for the binary setting;
+    'auto' selects 'upper' for binary and 'lower' for non-binary data;
+    input: reduced logits
+    """
+    if len(z.shape) < 2:
+        z = np.expand_dims(z, axis=1)
+
+    if bound=="auto":
+        # avg = np.mean(z)
+        # xxT = np.outer(z, z)
+        # avg_sq = np.mean(xxT)*n/(n-1) - np.mean(np.diag(xxT))/(n-1)
+        # # if False, we would get log of a negative value
+        # log_of_pos = 0 < (1 + (np.mean(z)**2 - np.var(z)/n)/(16*np.pi**2))
+        log_of_pos = True
+        upper_valid = (z.shape[1]==1) and log_of_pos
```

Comment:
`upper_valid` is now effectively only checking `z` is a vector rather than a matrix

Reply:
This code is also not from us